src/pool.rs

```diff
@@ -1071,4 +1077,103 @@ mod test {
             "Unexpected error: {claim_err}"
         );
     }
+
+    #[tokio::test]
+    async fn dropping_pool_closes_all_connections() {
+        async fn wait_for<F, Fut>(timeout: Duration, f: F) -> Result<(), Elapsed>
+        where
+            F: Fn() -> Fut,
+            Fut: Future<Output = bool>,
+        {
+            tokio::time::timeout(timeout, async move {
+                loop {
+                    if f().await {
+                        return;
+                    }
+                    tokio::time::sleep(Duration::from_millis(100)).await;
+                }
+            })
+            .await
+        }
+
+        setup_tracing_subscriber();
+
+        // Start a server that keeps connections open until the client closes
+        // them, and only keeps track of the counts.
+        let server_addr: SocketAddr = "127.0.0.1:0".parse().unwrap();
+        let n_active_conns = Arc::new(AtomicUsize::new(0));
+        let server_sock = TcpListener::bind(server_addr)
+            .await
+            .expect("bound localhost");
+        let server_addr = server_sock.local_addr().expect("got local_addr");
+        let server_handle = {
+            let n_active_conns = Arc::clone(&n_active_conns);
+            tokio::spawn(async move {
+                while let Ok((mut stream, _)) = server_sock.accept().await {
+                    n_active_conns.fetch_add(1, Ordering::Relaxed);
+                    let n_active_conns = n_active_conns.clone();
+                    tokio::spawn(async move {
+                        let mut buf = vec![0; 1024];
+                        loop {
+                            match stream.read(&mut buf).await {
+                                Ok(0) | Err(_) => {
+                                    n_active_conns.fetch_sub(1, Ordering::Relaxed);
+                                    return;
+                                }
+                                Ok(_) => continue,
+                            }
+                        }
+                    });
+                }
+            })
+        };
+
+        // Create the pool.
+        let spares_wanted = 4;
+        let resolver = Box::new(SingleHostResolver::new(server_addr));
+        let connector = Arc::new(TcpConnector {});
+        let pool = Pool::new(
+            resolver,
+            connector,
+            Policy {
+                spares_wanted,
+                set_config: SetConfig {
+                    // This is the default, but importantly, we want to test
+                    // that the connections get dropped before the next health
+                    // interval.
+                    health_interval: Duration::from_secs(30),
```

Comment:
I also made this explicit, just in case the default changes. I needed this to be a larger value than our timeouts, because I wanted to validate that connections are dropped **before** the periodic health check.